include/swift/AST/ModuleDependencies.h

```diff
@@ -447,55 +454,159 @@ class ModuleDependenciesCache {
+  ClangModuleDependenciesCacheImpl *getClangImpl() const { return clangImpl; }
+
+  /// Whether we have cached dependency information for the given module.
+  bool hasDependencies(StringRef moduleName,
+                       ModuleLookupSpecifics details) const;
+
+  /// Look for module dependencies for a module with the given name given
+  /// current search paths.
+  ///
+  /// \returns the cached result, or \c None if there is no cached entry.
+  Optional<ModuleDependencies>
+  findDependencies(StringRef moduleName, ModuleLookupSpecifics details) const;
+
+public:
+  /// Look for module dependencies for a module with the given name.
+  /// This method has a deliberately-obtuse name to indicate that it is not to
+  /// be used for general queries.
+  ///
+  /// \returns the cached result, or \c None if there is no cached entry.
+  Optional<ModuleDependenciesVector>
+  findAllDependenciesIrrespectiveOfSearchPaths(
+      StringRef moduleName, Optional<ModuleDependenciesKind> kind) const;
+
+  /// Record dependencies for the given module.
+  const ModuleDependencies* recordDependencies(StringRef moduleName,
+                                               ModuleDependencies dependencies);
+
+  /// Update stored dependencies for the given module.
+  const ModuleDependencies* updateDependencies(ModuleDependencyID moduleID,
+                                               ModuleDependencies dependencies);
+
+  /// Reference the list of all module dependencies.
+  const std::vector<ModuleDependencyID> &getAllModules() const {
+    return AllModules;
+  }
+};
+
+/// This "local" dependencies cache persists only for the duration of a given scanning action,
+/// and wraps an instance of a `GlobalModuleDependenciesCache` which may carry cached
+/// scanning information from prior scanning actions. This cache maintains a store of references
+/// to all dependencies found within the current scanning action (with their values stored in the
+/// global Cache), since these do not require clients to disambiguate them with search paths.
+class ModuleDependenciesCache {
+private:
+  GlobalModuleDependenciesCache &globalCache;
+
+  /// Dependencies for Textual Swift modules that have already been computed.
+  /// References to data in `globalCache`.
+  llvm::StringMap<const ModuleDependencies*> SwiftTextualModuleDependencies;
+
+  /// Dependencies for Binary Swift modules that have already been computed.
+  /// References to data in `globalCache`.
+  llvm::StringMap<const ModuleDependencies*> SwiftBinaryModuleDependencies;
+
+  /// Dependencies for Swift placeholder dependency modules that have already
+  /// been computed.
+  /// References to data in `globalCache`.
+  llvm::StringMap<const ModuleDependencies*> SwiftPlaceholderModuleDependencies;
+
+  /// Dependencies for Clang modules that have already been computed.
+  /// References to data in `globalCache`.
+  llvm::StringMap<const ModuleDependencies*> ClangModuleDependencies;
```

Comment:
Can we define one dictionary from dependency kind to `StringMap` instead of four StrignMaps?

Reply:
Good idea. Done.